### PR TITLE
Simplify kube-apiserver patches

### PR DESCRIPTION
**Pull request by deads2k**
use the config poststarthooks to further simplify the kube-apiserver patches.

/assign @p0lyn0mial

### CODE REVIEW

**Comment by mfojtik** [comment]
The gcp failure was interesting: 

```
I1120 22:38:53.916977       1 log.go:172] http2: panic serving [::1]:53440: open /etc/kubernetes/secrets/etcd-client.crt: no such file or directory
goroutine 455847 [running]:
github.com/openshift/origin/vendor/k8s.io/apiserver/pkg/server/filters.(*timeoutHandler).ServeHTTP.func1.1(0xc01652c1e0)
```

**Comment by stlaz** [comment]
As for the above error, it seems that the goroutine is being restarted in a loop which causes about 21k lines of panics in the bootstrap-control-plane kube-apiserver logs

File: vendor/k8s.io/kubernetes/cmd/kube-apiserver/app/aggregator.go [hunk 1]
```diff
@@ -104,6 +104,8 @@ func createAggregatorConfig(
 			EnableAggregatedDiscoveryTimeout: utilfeature.DefaultFeatureGate.Enabled(kubefeatures.EnableAggregatedDiscoveryTimeout),
 		},
 	}
+	// we need to clear the poststarthooks so we don't add them multiple times to all the servers (that fails)
+	aggregatorConfig.GenericConfig.PostStartHooks = map[string]genericapiserver.PostStartHookConfigEntry{}
```
Comment (sttts) [inline vendor/k8s.io/kubernetes/cmd/kube-apiserver/app/aggregator.go]: is there an upstream PR already? Please rename the commit.

File: vendor/k8s.io/kubernetes/openshift-kube-apiserver/openshiftkubeapiserver/patch_handlerchain.go [hunk 2]
```diff
@@ -12,10 +12,23 @@ import (
...
+func BuildHandlerChain(consolePublicURL string, oauthMetadataFile string) (func(apiHandler http.Handler, kc *genericapiserver.Config) http.Handler, error) {
+	// load the oauthmetadata when we can return an error
+	oAuthMetadata := []byte{}
+	if len(oauthMetadataFile) > 0 {
+		var err error
+		oAuthMetadata, err = loadOAuthMetadataFile(oauthMetadataFile)
+		if err != nil {
+			return nil, err
+		}
+	}
```
Comment (sttts) [inline vendor/k8s.io/kubernetes/openshift-kube-apiserver/openshiftkubeapiserver/patch_handlerchain.go]: strange to have file loading in a handler chain builder. Why not pass in a `[]byte`?

File: vendor/k8s.io/kubernetes/cmd/kube-apiserver/app/server.go [hunk 3]
```diff
@@ -30,7 +30,8 @@ import (
 	"strings"
 	"time"
 
-	"k8s.io/kubernetes/openshift-kube-apiserver/configdefault"
+	"k8s.io/kubernetes/openshift-kube-apiserver/admission/admissionenablement"
+
```
Comment (sttts) [inline vendor/k8s.io/kubernetes/cmd/kube-apiserver/app/server.go]: no empty line

File: vendor/k8s.io/kubernetes/pkg/kubeapiserver/authenticator/patch_authenticator.go [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+	"k8s.io/apiserver/pkg/authentication/group"
+	genericapiserver "k8s.io/apiserver/pkg/server"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/keyutil"
+	"k8s.io/kubernetes/openshift-kube-apiserver/admission/authorization/restrictusers/usercache"
+	oauthvalidation "k8s.io/kubernetes/openshift-kube-apiserver/admission/customresourcevalidation/oauth"
+	"k8s.io/kubernetes/openshift-kube-apiserver/authentication/oauth"
+	"k8s.io/kubernetes/openshift-kube-apiserver/enablement"
+	"k8s.io/kubernetes/pkg/serviceaccount"
```
Comment (stlaz) [inline vendor/k8s.io/kubernetes/pkg/kubeapiserver/authenticator/patch_authenticator.go]: nit: split imports into openshift/kube groups

File: vendor/k8s.io/kubernetes/pkg/kubeapiserver/authenticator/patch_authenticator.go [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+	oauthInformer := oauthinformer.NewSharedInformerFactory(oauthClient, defaultInformerResyncPeriod)
+	userInformer := userinformer.NewSharedInformerFactory(userClient, defaultInformerResyncPeriod)
+	if err := userInformer.User().V1().Groups().Informer().AddIndexers(cache.Indexers{
+		usercache.ByUserIndexName: usercache.ByUserIndexKeys,
+	}); err != nil {
+		panic(err)
+	}
+
+	if enablement.IsOpenShift() {
```
Comment (stlaz) [inline vendor/k8s.io/kubernetes/pkg/kubeapiserver/authenticator/patch_authenticator.go]: We don't need to do any of the above if this is false, do we?

File: vendor/k8s.io/kubernetes/pkg/kubeapiserver/authenticator/patch_authenticator.go (hunk 4 above)
Comment (deads2k) [reply vendor/k8s.io/kubernetes/pkg/kubeapiserver/authenticator/patch_authenticator.go]: > nit: split imports into openshift/kube groups

this is our file.  I really see no benefit to keeping them separate now that we have auto-importing capability (didn't exist 5 years ago)

Comment (stlaz) [reply vendor/k8s.io/kubernetes/pkg/kubeapiserver/authenticator/patch_authenticator.go]: IMO it's a good habit to keep separate import contexts visibly separate, like syslibs/imports from elsewhere/imports from us

**Comment by stlaz** [comment]
/hold

Kube request header auth was put higher in the eval stack than the token auth, which causes the `OAuth-server has the correct token and certificate fallback semantics` to fail.

We probably still want our token auth to precede kube req headers?

**Comment by deads2k** [comment]
/hold

We have a fun pickle regarding `--token` now.  I need to think about this.

**Comment by deads2k** [comment]
added release note https://github.com/openshift/openshift-docs/issues/18426 after clearing it with Clayton.

/hold cancel

**Comment by stlaz** [comment]
/lgtm
The release note looks good and we need this change for further work.

**Pull request merged**